drivers/eeprom/eeprom_simulator.c

```diff
@@ -0,0 +1,309 @@
+STATS_NAME(eeprom_sim_stats, bytes_written)
+STATS_NAME(eeprom_sim_stats, eeprom_read_calls)
+STATS_NAME(eeprom_sim_stats, eeprom_read_time_us)
+STATS_NAME(eeprom_sim_stats, eeprom_write_calls)
+STATS_NAME(eeprom_sim_stats, eeprom_write_time_us)
+STATS_NAME_END(eeprom_sim_stats);
+
+/* simulator dynamic thresholds */
+STATS_SECT_START(eeprom_sim_thresholds)
+STATS_SECT_ENTRY32(max_write_calls)
+STATS_SECT_ENTRY32(max_len)
+STATS_SECT_END;
+
+STATS_SECT_DECL(eeprom_sim_thresholds) eeprom_sim_thresholds;
+STATS_NAME_START(eeprom_sim_thresholds)
+STATS_NAME(eeprom_sim_thresholds, max_write_calls)
+STATS_NAME(eeprom_sim_thresholds, max_len)
+STATS_NAME_END(eeprom_sim_thresholds);
+
+#ifdef CONFIG_ARCH_POSIX
+static u8_t *mock_eeprom;
+static int eeprom_fd = -1;
+static const char *eeprom_file_path;
+static const char default_eeprom_file_path[] = "eeprom.bin";
+#else
+static u8_t mock_eeprom[DT_INST_0_ZEPHYR_SIM_EEPROM_SIZE];
+#endif /* CONFIG_ARCH_POSIX */
+
+static int eeprom_range_is_valid(struct device *dev, off_t offset, size_t len)
+{
+	const struct eeprom_sim_config *config = DEV_CONFIG(dev);
+
+	if ((offset + len) < config->size) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static int eeprom_sim_read(struct device *dev, const off_t offset, void *data,
+			  const size_t len)
+{
+	if (!len) {
+		return 0;
+	}
+
+	if (!eeprom_range_is_valid(dev, offset, len)) {
+		LOG_WRN("attempt to read past device boundary");
+		return -EINVAL;
+	}
+
+	SYNC_LOCK();
+
+	STATS_INC(eeprom_sim_stats, eeprom_read_calls);
+	memcpy(data, EEPROM(offset), len);
+	STATS_INCN(eeprom_sim_stats, bytes_read, len);
+
+	SYNC_UNLOCK();
+
+#ifdef CONFIG_EEPROM_SIMULATOR_SIMULATE_TIMING
+	k_busy_wait(CONFIG_EEPROM_SIMULATOR_MIN_READ_TIME_US);
+	STATS_INCN(eeprom_sim_stats, eeprom_read_time_us,
+		   CONFIG_EEPROM_SIMULATOR_MIN_READ_TIME_US);
+#endif
+
+	return 0;
+}
+
+static int eeprom_sim_write(struct device *dev, const off_t offset,
+			   const void *data, const size_t len)
+{
+	const struct eeprom_sim_config *config = DEV_CONFIG(dev);
+
+	if (config->readonly) {
+		LOG_WRN("attempt to write to read-only device");
+		return -EACCES;
+	}
+
+	if (!len) {
+		return 0;
+	}
+
+	if (!eeprom_range_is_valid(dev, offset, len)) {
+		LOG_WRN("attempt to write past device boundary");
+		return -EINVAL;
+	}
+
+	SYNC_LOCK();
+
+	STATS_INC(eeprom_sim_stats, eeprom_write_calls);
+
+	bool data_part_ignored = false;
+
+	if (eeprom_sim_thresholds.max_write_calls != 0) {
+		if (eeprom_sim_stats.eeprom_write_calls >
+			eeprom_sim_thresholds.max_write_calls) {
+			return 0;
+		} else if (eeprom_sim_stats.eeprom_write_calls ==
+				eeprom_sim_thresholds.max_write_calls) {
+			if (eeprom_sim_thresholds.max_len == 0) {
+				return 0;
+			}
+
+			data_part_ignored = true;
+		}
+	}
+
+	for (u32_t i = 0; i < len; i++) {
+		if (data_part_ignored) {
+			if (i >= eeprom_sim_thresholds.max_len) {
+				return 0;
+			}
+		}
+
+		*(EEPROM(offset + i)) = *((u8_t *)data + i);
+	}
```

Comment:
What is the purpose of iterating through all the bytes? Limiting the number of bytes written overall based on `max_len`? If so, wouldn't it be sufficient to cap `len` to `max_len` and use `memcpy()` or am I missing something?

Reply:
@brix, this allows you to simulate write errors, anything that is beyond the max_len is ignored. An example of how this can be used can be found in tests/subsys/fs/nvs.

Reply:
@Laczen I get that much. But couldn't you just do something like `MAX(len, max_len)` and do a `memcpy()`?

Reply:
@brix, updated. Made me realize that there was something wrong with the lock, corrected.